pubtools/exodus/hooks/pulp.py

```diff
@@ -0,0 +1,165 @@
+        session = self._new_session()
+        resp = session.request(**kwargs)
+        self._unpack_response(resp)
+        return resp
+
+    def _check_cert(self):
+        """Issue request to exodus-gw to identify permissions."""
+
+        if not self._gw_crt:
+            LOG.debug("EXODUS_GW_CERT not found, authentication may fail")
+
+        if not self._gw_key:
+            LOG.debug("EXODUS_GW_KEY not found, authentication may fail")
+
+        auth_url = urljoin(self._gw_url, "/whoami")
+        resp = self._do_request(method="GET", url=auth_url)
+        context = resp.json()
+
+        for user_type, ident in (
+            ("client", "serviceAccountId"),
+            ("user", "internalUsername"),
+        ):
+            typed_ctx = context[user_type]
+            if typed_ctx["authenticated"]:
+                LOG.debug(
+                    "You are authenticated as %s %s with roles: %s",
+                    user_type,
+                    typed_ctx[ident],
+                    [str(role) for role in typed_ctx["roles"]],
+                )
+                break
+        else:
+            LOG.debug("You are not authenticated with exodus-gw.")
+
+    @classmethod
+    def _unpack_response(cls, response):
+        """Raise if 'response' was not successful.
+
+        This is the same as response.raise_for_status(), merely wrapping it
+        to ensure the body is logged when possible."""
+
+        try:
+            response.raise_for_status()
+        except Exception as outer:
+            try:
+                body = response.json()
+            except:
+                raise outer
+
+            LOG.error("Unsuccessful response from exodus-gw: %s", body)
+            raise
+
+    @backoff.on_predicate(
+        wait_gen=backoff.expo,
+        predicate=lambda task: task["state"] not in ("FAILED", "COMPLETE"),
+        max_time=1800,
+    )
+    def _poll_commit_completion(self, commit):
+        """Issues request(s) to exodus-gw to get the task state, returning once
+        it either is FAILED or COMPLETED."""
+        task_url = urljoin(self._gw_url, commit["links"]["self"])
+
+        resp = self._do_request(method="GET", url=task_url)
+        task = resp.json()
+
+        LOG.info("Task state: %s", task["state"])
+
+        return task
+
+    @hookimpl
+    def pulp_repository_pre_publish(self, to_publish):
+        """Invoked before a Pulp repository is published.
+
+        This implementation adds to each config the --exodus-publish argument,
+        attaching the repository to an Exodus gateway publish.
+
+        Args:
+            to_publish (list[tuple[:class:`~pubtools.pulplib.Distributor`, dict]])
+                List of Distributor-config pairs intended for publishing.
+
+        Returns:
+            to_publish (list[tuple[:class:`~pubtools.pulplib.Distributor`, dict]])
+                List of potentially adjusted Distributor-config pairs to publish.
+        """
+
+        if not self._publish:
```

Comment:
This probably needs some locking to create the publish only once.  It is unclear if pulp_repository_pre_publish can be invoked from different threads simultaneously.